# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
# Load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

os.getcwd()

'/Users/chris/code/Springboard/GuidedCapstone'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv('./data/step3_output.csv', index_col=0)
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         0  
1               90.0             0.0         0  
2              152.0            30.0         0  
3              122.0             0.0         1  
4              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
# Select object type columns
state_indicators = pd.get_dummies(df['state'])
# Drop state column
df.drop(columns='state', inplace=True)
# Concatenate state indicators
df = pd.concat([df, state_indicators], axis=1)

In [4]:
df.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend','clusters'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
x_scaled = scaler.transform(x)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1 - Linear Regression

In [7]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import explained_variance_score, mean_absolute_error

In [8]:
lr_model = LinearRegression()
lr_model.fit(x_train, y_train)
lr_model.score(x_test, y_test)

0.9407146952371528

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lr_model.predict(x_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [11]:
explained_variance_score(y_test, y_pred)

0.9407854220196027

In [12]:
mean_absolute_error(y_test, y_pred)

4.913713440313831

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
lr_model.intercept_

64.09934644291931

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef_df = pd.DataFrame(abs(lr_model.coef_), x.columns, columns=['Coefficient'])
coef_df.sort_values(by='Coefficient', ascending=False).head(10)

Coefficient
total_chairs  6.384564e+13
fastQuads     2.420403e+13
surface       2.267736e+13
double        1.998413e+13
New York      1.859552e+13
triple        1.782722e+13
Michigan      1.754912e+13
Colorado      1.546179e+13
California    1.513081e+13
quad          1.444830e+13

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

In [15]:
# Import dataset again to remove dummy features created at the beginning of this notebook
df = pd.read_csv('./data/step3_output.csv', index_col=0)

In [16]:
# Drop unneeded features
x = df.drop(['Name','AdultWeekend','clusters','state'], axis=1)
# Define actionable trait
y = df['AdultWeekend']

In [17]:
# Scaling
scaler = preprocessing.StandardScaler().fit(x)
x_scaled = scaler.transform(x)

In [18]:
# Get the 1-dimensional flattened array of our response variable
y = y.ravel()

# Generate training data and tests
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

In [19]:
# Fit and score model
lr_model = LinearRegression()
lr_model.fit(x_train, y_train)
lr_model.score(x_test, y_test)

0.9280431232590022

In [20]:
# Generate actionable trait predictions using feature tests
y_pred = lr_model.predict(x_test)

In [21]:
# Generate the explained variance score (EVS)
explained_variance_score(y_test, y_pred)

0.9282915219558896

In [22]:
# Generate the Mean Absolute Error
mean_absolute_error(y_test, y_pred)

5.405981008721084

In [23]:
# Intercept
lr_model.intercept_

64.0632540314099

In [24]:
# Top feature coefficients
coef_df = pd.DataFrame(abs(lr_model.coef_), x.columns, columns=['Coefficient'])
coef_df.sort_values(by='Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday        20.119216
averageSnowfall      1.773474
vertical_drop        1.622903
quad                 1.469428
triple               1.352074
Runs                 1.188930
surface              1.175385
daysOpenLastYear     1.153050
summit_elev          1.030203
fastQuads            0.692876

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [25]:
# Import dataset again to remove dummy features created at the beginning of this notebook
df = pd.read_csv('./data/step3_output.csv', index_col=0)

In [26]:
# Drop unneeded features
# NOTE: 'base_elev' was dropped in Step 3
x = df.drop(['Name','AdultWeekend','clusters','state','summit_elev'], axis=1)
# Define actionable trait
y = df['AdultWeekend']

In [27]:
# Scaling
scaler = preprocessing.StandardScaler().fit(x)
x_scaled = scaler.transform(x)

In [28]:
# Get the 1-dimensional flattened array of our response variable
y = y.ravel()

# Generate training data and tests
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

In [29]:
# Fit and score model
lr_model = LinearRegression()
lr_model.fit(x_train, y_train)
lr_model.score(x_test, y_test)

0.9240132105868615

In [30]:
# Generate actionable trait predictions using feature tests
y_pred = lr_model.predict(x_test)

In [31]:
# Generate the explained variance score (EVS)
explained_variance_score(y_test, y_pred)

0.9242631433903842

In [32]:
# Generate the Mean Absolute Error
mean_absolute_error(y_test, y_pred)

5.556374146784197

In [33]:
# Intercept
lr_model.intercept_

64.06293474346353

In [34]:
# Top feature coefficients
coef_df = pd.DataFrame(abs(lr_model.coef_), x.columns, columns=['Coefficient'])
coef_df.sort_values(by='Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday        20.067543
averageSnowfall      2.169304
quad                 1.550772
triple               1.336327
Runs                 1.305905
surface              1.206186
vertical_drop        1.152284
daysOpenLastYear     0.979620
fastQuads            0.774892
total_chairs         0.711019

#### Model 4 - Random Forest

In [35]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

In [36]:
# Import dataset again for a fresh start
df = pd.read_csv('./data/step3_output.csv', index_col=0)

In [37]:
# Drop unneeded features
# NOTE: 'base_elev' was dropped in Step 3
x = df.drop(['Name','AdultWeekend','clusters','state','summit_elev'], axis=1)
# Define actionable trait
y = df['AdultWeekend']

In [38]:
# Scaling
scaler = preprocessing.StandardScaler().fit(x)
x_scaled = scaler.transform(x)

In [39]:
# Get the 1-dimensional flattened array of our response variable
y = y.ravel()

# Generate training data and tests
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.25, random_state=1)

In [40]:
# Fit model
rfr_model = RandomForestRegressor()
rfr_model.fit(x_train, y_train)
rfr_model.score(x_test, y_test)

0.9012165724104559

In [41]:
# Make predictions
y_pred = rfr_model.predict(x_test)

In [42]:
coef_df = pd.DataFrame(abs(rfr_model.feature_importances_), x.columns, columns=['Feature Importances'])
coef_df.sort_values(by='Feature Importances', ascending=False).head(10)

Feature Importances
AdultWeekday                  0.847383
Snow Making_ac                0.020283
Runs                          0.013389
total_chairs                  0.012852
vertical_drop                 0.011136
yearsOpen                     0.010350
SkiableTerrain_ac             0.009940
LongestRun_mi                 0.008928
averageSnowfall               0.008655
daysOpenLastYear              0.008640

In [43]:
explained_variance_score(y_test, y_pred)

0.9012239487520287

In [44]:
mean_absolute_error(y_test, y_pred)

6.036579967180553

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9407854220196027 | 4.913713440313831 |-|
| Model 2. | 0.9282915219558896| 5.405981008721084 |'state'|
| Model 3. | 0.9242631433903842 | 5.556374146784197 |'state','summit_elev','base_elev'|
| Model 4. | 0.9049261933520774 | 5.917000729800926 |'state','summit_elev','base_elev'|

Model Selection:

**Model 3 - Linear Regression with `state`, `summit_elev`, and `base_elv` features dropped.**